baybe/recommenders/base.py

```diff
@@ -212,15 +194,20 @@ def _select_candidates_and_recommend(
             NotEnoughPointsLeftError: If there are fewer points left for potential
                 recommendation than requested.
         """
+        is_hybrid_space = searchspace.type is SearchSpaceType.HYBRID
+
         # Get discrete candidates
-        _, candidates_comp = subspace_discrete.get_candidates(
-            allow_repeated_recommendations=self.allow_repeated_recommendations,
-            allow_recommending_already_measured=self.allow_recommending_already_measured,
+        _, candidates_comp = searchspace.discrete.get_candidates(
+            allow_repeated_recommendations=True
+            if is_hybrid_space
+            else self.allow_repeated_recommendations,
+            allow_recommending_already_measured=True
+            if is_hybrid_space
+            else self.allow_recommending_already_measured,
         )
 
         # Check if enough candidates are left
-        # TODO [15917]: This check is not perfectly correct.
-        if len(candidates_comp) < batch_size:
+        if (not is_hybrid_space) and (len(candidates_comp) < batch_size):
```

Comment:
Why did you remove the TODO? Isn't the check basically still the same?

Reply:
I wasnt sure whats not correct here hence assumed this todo is old, do you know?

One thing that comes to mind is: if len(candidates) < batch size one could still (trivially) recommend the remaining points. But that is not really conform with the function returning exactly `batch size` points as per current definition and assumption

Reply:
I also think that this was the original reason for this TODO, but imo, the check and the corresponding error message is correct. So we can remove it here or add a small comment in the code about "There might in general still be some points but not enough". Do as you like and then resolve :)

Reply:
yes i know, clearly it hasn't helped me in remembering it anyway :) which is of course on me

My point: it will happen again to others (or me :D )

Reply:
Hey guys, sorry, don't want to disturb, just saw this on my phone (really need to disable GH push messages now :D). But the TODO was not about having too few points. It's about the fact that technically checking the len of the "candidates" is not enough. Example: get_candidates gives me 2 points that are left in the searchspace, and I request 5. If "allow_repeated_recommendations" is `True`, I can still return 5 recommendations without problem by choosing them "mit zurücklegen"

Reply:
so if the additional three recommendations are copies of the remaining two and recommended solely for the purpose of having a total of 5 - whats the point of that?  just to satisfy the docstring? Seems like a no-usecase for me

so imo this check is complete and the TODO obsolete

Keep in mind that this will only happen anyway if someone measures all combinations... which is not a use case of baybe

The error we throw is even suable in simulations module to detect that the simulation cannot cotinue

but to unblock us Ill restore the old TODO and merge

Reply:
I think what you suggested (just restore and continue) sounds reasonable. And while not strictly necessary in most use cases, I wouldn't say it's irrelevant. Especially in high noise situations, it's a valid question to ask which option to measure further. By saying this is irrelevant, you basically give the finger to the entire bandits community 😄

Reply:
not so fast, I never said that the idea of duplicate recommendations per se is irrelevant, I agree that in the situation of noisy measurements an algorithm recommending duplicates due to noise-awareness would make a lot of sense

However, that is not the situation this TODO refers to, and that is not why this code would recommend duplicates. It would only do so to fill in the amount so it recommends to the batch size -> this has nothing to do with the above -> and imo doesnt make sense -> hence I think the todo is obsolete (your idea is better captured in a work item and not this todo)

btw the TODO references a number which I thought could be a work item but I wasnt able to find it

Reply:
resolving because it was `reintroduced`